crates/flux-desugar/src/desugar.rs

```diff
@@ -1581,21 +1633,65 @@ fn index_sort(
     resolver_output: &ResolverOutput,
     bty: &surface::BaseTy,
 ) -> Option<fhir::Sort> {
-    // CODESYNC(sort-of, 4) sorts should be given consistently
+    // CODESYNC(sort-of, 3) sorts should be given consistently
     match &bty.kind {
-        surface::BaseTyKind::Path(path) => {
-            let res = resolver_output.path_res_map[&path.node_id];
-            genv.sort_of_res(res)
-        }
+        surface::BaseTyKind::Path(path) => sort_of_surface_path(genv, resolver_output, path),
         surface::BaseTyKind::Slice(_) => Some(fhir::Sort::Int),
     }
 }
 
-fn as_tuple<'a>(genv: &'a GlobalEnv, sort: &'a fhir::Sort) -> &'a [fhir::Sort] {
-    if let fhir::Sort::Record(def_id) = sort {
-        genv.index_sorts_of(*def_id)
+fn sort_of_surface_path(
+    genv: &GlobalEnv,
+    resolver_output: &ResolverOutput,
+    path: &surface::Path,
+) -> Option<fhir::Sort> {
+    // CODESYNC(sort-of-path, 2) sorts should be given consistently
+    let res = resolver_output.path_res_map[&path.node_id];
+
+    match res {
+        fhir::Res::PrimTy(PrimTy::Int(_) | PrimTy::Uint(_)) => Some(fhir::Sort::Int),
+        fhir::Res::PrimTy(PrimTy::Bool) => Some(fhir::Sort::Bool),
+        fhir::Res::PrimTy(PrimTy::Float(..) | PrimTy::Str | PrimTy::Char) => Some(fhir::Sort::Unit),
+        fhir::Res::Def(DefKind::TyAlias { .. } | DefKind::Enum | DefKind::Struct, def_id) => {
+            // TODO: duplication with sort_of_path
+            let mut sort_args = vec![];
+            if let Some(generics) = genv.map().get_generics(def_id) {
```

Comment:
This code is assuming that external types don't have polymorphic sorts. That's wrong and we should use `rty::Generics` instead. However, I'm ok with leaving it like this for now. My intention is to get rid of this function eventually as the logic is too subtle to have it duplicated.

Reply:
Another though. There's an implicit _synchronization_ here that maps the generic arguments of the path to the generic sorts of the index. Maybe we should make that mapping explicit. For example, instead of iterating over the generic args of the path, we could iterate over `RefinedBy::sort_params`. If we do that, only `RefinedBy` needs to be generic of the order/mapping (this makes https://github.com/flux-rs/flux/pull/542#discussion_r1376787104 obsolete)